Add pulp handler and dnf support.

jortel opened the pull request:
http://projects.theforeman.org/issues/21904

Summary:
- Add (stripped down) pulp agent handler framework.  This is needed to provide integration between pulp and the agent.
- Add the DNF library[1] (wrapper). _New_.
- Add the YUM library[1] (wrapper).
- Add the RPM and errata handlers[1]. 
- Changes to katelloagent (plugin):
  - Installed (and loaded from) site-packages.
  - The Conduit removed.  This provided support for progress reporting and cancellation.  Neither used by katello.
  - Use the forking dispatch model.  This causes goferd to dispatch each RMI call to the plugin in a new child process.  This is needed because although the dnf lib is being initialized and closed correctly, bad things happen when using the lib twice in the same process.  This also ensure that dnf (and yum) memory (and file descriptor) leaks cannot impact goferd (daemon). 
- I took the liberty of organizing things into python packages.  The `katello.agent.goferd` packages contains the `plugin`.   The `katello.agent.pulp` package provides Pulp integration.

**[1]** Taken from pulp, stripped down, refactored a bit and ported to python 2.7/3.x.

Flows and abstraction layers:
```
Pulp (request) => goferd => katello <plugin> => Dispatcher => Handler => LibAPI => (DNF|YUM)
            <================ DispatchReport <= HandlerReport <= TransactionReport
```
---
*** THIS IS A WORK IN PROGRESS  ***

- Needs to be tested running in goferd.
- Needs to be tested w/ python3.
- The dnf support needs to log packages installed/updated/removed.
---
To run on python3, this will need gofer 2.12+.

The python3 compatible version (likely 2.12) is not packaged yet so for testing, it will need to be installed from source.

Deps:
- pam
- python3-six
- python3-future
- python3-qpid-proton

Steps:
```
# mkdir git
# git clone https://github.com/jortel/gofer.git
# git checkout python3-compat
# pushd gofer/src
# python3 setup.py install --no-compile --install-purelib /usr/lib/python3.6/site-packages
# popd
# mkdir -p /etc/gofer/plugins
# cp etc/gofer/agent.conf /etc/gofer
# cp docs/man/man1/goferd.1 /usr/share/man/man1
# vi bin/goferd

- #! /usr/bin/env python
+ #! /usr/bin/env python3

```
Running.  See: `man goferd`.
```
# bin/goferd 
# ps -ef |grep goferd
root      4292     1  0 15:16 ?        00:00:00 python bin/goferd
```
---

The plugin:
```
# cd git
# git clone https://github.com/jortel/katello-host-tools.git
# git checkout pulp-handler
# pushd src
# python3 setup.py install --no-compile --install-purelib /usr/lib/python3.6/site-packages
# popd
# cp etc/gofer/plugins/katello.conf /etc/gofer/plugins/
```

Restart goferd.
#

jturel commented on src/katello/agent/goferd/plugin.py:
```diff
@@ -15,27 +15,35 @@
 from gofer.config import Config
 
-import enabled_repos_upload
+try:
```
This PR needs to be rebased from master and this try can then be replaced to get this plugin loaded properly:

```
+from katello.enabled_report import EnabledReport
+from katello.repos import upload_enabled_repos_report
```

send_enabled_report will need to be updated:

```
 def send_enabled_report(path=REPOSITORY_PATH):
-    report = enabled_repos_upload.EnabledReport(path)
-    enabled_repos_upload.upload_enabled_repos_report(report)
+    report = EnabledReport(path)
+    upload_enabled_repos_report(report)

```

jturel commented on src/katello/agent/goferd/plugin.py:
```diff
@@ -127,14 +134,16 @@ def certificate_changed(path):
     enabled_repos_upload.upload_enabled_repos_report(report)
 
+
 def update_settings():
```
update_settings needs a tweak for PY3:

```
-    if rhsm_conf['rhsm'].has_key('ca_cert_dir'):
+    if 'ca_cert_dir' in rhsm_conf['rhsm']:

```

jturel commented on src/katello/agent/goferd/plugin.py:
```diff
@@ -15,27 +15,35 @@
 import sys
-import httplib
+
+if PY2:
```
according to python-six docs instead of loading httplibs conditionally:

`from six import http_client as http`

didn't test it though :)

jturel commented on src/katello/agent/goferd/plugin.py:
```diff
@@ -15,27 +15,35 @@
 from time import sleep
 from logging import getLogger
 from subprocess import Popen
 
```
I was getting a stack trace from `path_monitor.add(REPOSITORY_PATH, send_enabled_report)` which I commented out locally to test. It looks like something will need to be updated for PY3 but not sure just what needs to happen there:

```
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [INFO][MainThread] gofer.agent.plugin:686 - plugin:katello loaded using: katello.agent.goferd.plugin
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 - plugin:katello, import failed
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 - Traceback (most recent call last):
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/agent/plugin.py", line 694, in _load
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     plugin.load()
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/common.py", line 312, in sfn
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     return fn(*args, **kwargs)
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/agent/plugin.py", line 440, in load
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     self.delegate.loaded()
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/agent/decorator.py", line 89, in loaded
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     fn()
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/katello/agent/goferd/plugin.py", line 86, in init_plugin
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     path_monitor.add(path, certificate_changed)
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/common.py", line 312, in sfn
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     return fn(*args, **kwargs)
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/pmon.py", line 167, in add
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     self._paths.add(Tracker(path, target))
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/pmon.py", line 108, in __init__
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     self.digest = digest(path)
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -   File "/usr/lib/python3.6/site-packages/gofer/pmon.py", line 68, in digest
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 -     _hash.update(s)
Apr 06 18:45:45 fedora27.strangeways.example.com goferd[13547]: [ERROR][MainThread] gofer.agent.plugin:697 - TypeError: Unicode-objects must be encoded before hashing
```

jortel replied:
Fixed and bulit in copr.  Because the version is based on the commit hash, you may need to reinstall.

jturel commented on src/katello/agent/pulp/libyum.py:
```diff
@@ -0,0 +1,361 @@
+            for pattern in patterns:
+                try:
+                    lib.install(pattern=str(pattern))
+                except InstallError, caught:
```
`InstallError as caught`

jturel commented:
I tested this with gofer from the copr repo you provided. However, I did need to install python3-future by hand (iirc) so there may be a missing dep for that RPM.

jortel commented:
I had updated the code to discontinue using `python-future` and removed the dep since python-six already has all I need.  Looks like I missed a spot:
```
past.builtins import basestring
```
Will fix.

jturel commented:
I was setting up another box to test py2 and yum when suddenly..

(I built from HEAD of python3-compat)

```
Traceback (most recent call last):
  File "./bin/goferd", line 17, in <module>
    from gofer.agent.main import main
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/agent/main.py", line 34, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/agent/plugin.py", line 33, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/agent/rmi.py", line 19, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/agent/builtin.py", line 20, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/decorators.py", line 19, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/rmi/model/__init__.py", line 20, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/rmi/model/fork.py", line 16, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/rmi/model/parent.py", line 21, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/rmi/context.py", line 19, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/__init__.py", line 20, in <module>
  File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/auth.py", line 18, in <module>
ImportError: No module named builtins
```

More instances?
```
bin/gofer
17:from builtins import str

src/gofer/messaging/auth.py
18:from builtins import bytes

src/build/lib/gofer/messaging/auth.py
18:from builtins import bytes

src/build/scripts-2.7/gofer
17:from builtins import str
```

jortel commented:
Interesting.  That should have been caught by unit test and wondered why until I noticed the unit test runner script was installing _python-future_ in the vitrual-env when running them.

There are a few things that `future` does better than `six` so going back to using it as well.  Building in copr now.

jturel commented:
I guess it can't read inside the egg archive? Guess this wouldn't show up on my other box where I installed via RPM heh

```
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [INFO][Thread-1] gofer.rmi.store:108 - Using: /var/lib/gofer/messaging/pending/katello                                                                                                                                      
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [INFO][MainThread] gofer.agent.plugin:686 - plugin:katello loaded using: katello.agent.goferd.plugin                                                                                                                        
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [INFO][MainThread] rhsm.connection:821 - Connection built: host=centos7-katello-3-5.strangeways.example.com port=443 handler=/rhsm auth=identity_cert ca_dir=/etc/rhsm/ca/ insecure=True                                    
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [INFO][MainThread] rhsm.connection:551 - Response: status=200, request="GET /rhsm/consumers/bd7f8439-842d-439e-bce0-61bec6feeb98"                                                                                           
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [INFO][MainThread] katello.agent.goferd.plugin:163 - Using /etc/rhsm/ca/katello-default-ca.pem as the ca cert for qpid connection                                                                                           
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 - [Errno 20] Not a directory: '/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter'                                                          
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 - Traceback (most recent call last):                                                                                                                                     
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter/model.py", line 40, in _fn                                                     
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -     return fn(*args, **keywords)                                                                                                                                       
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter/model.py", line 360, in declare                                                
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -     adapter = Adapter.find(url)                                                                                                                                        
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter/factory.py", line 164, in find                                                 
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -     _list, catalog = Adapter.loader.load()                                                                                                                             
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter/factory.py", line 117, in load                                                 
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -     _list, catalog = Loader._load()                                                                                                                                    
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter/factory.py", line 92, in _load                                                 
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 -     for name in sorted(os.listdir(_dir)):                                                                                                                              
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.messaging.adapter.model:44 - OSError: [Errno 20] Not a directory: '/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter'                                                 
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 - 82daa44c-77f9-47e9-825a-6cc7fcdb25d5  
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 - Traceback (most recent call last):    
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/threadpool.py", line 98, in __call__                                                                         
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -     self.fn(*self.args, **self.kwargs)                                                                                                                                             
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/agent/plugin.py", line 53, in call                                                                           
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -     fn(plugin)                        
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/common.py", line 315, in sfn                                                                                 
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -     return fn(*args, **kwargs)        
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/agent/plugin.py", line 367, in attach                                                                        
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -     model.setup()                     
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/common.py", line 349, in _fn                                                                                 
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -     return fn(*args, **kwargs)        
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/agent/plugin.py", line 561, in setup                                                                         
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -     queue.declare(url)                
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -   File "/usr/lib/python2.7/site-packages/gofer-2.10.0-py2.7.egg/gofer/messaging/adapter/model.py", line 45, in _fn                                                                 
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 -     raise ModelError(*e.args)         
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [ERROR][worker-0] gofer.threadpool:100 - ModelError: (20, 'Not a directory')   
Apr 10 22:45:56 centos7.strangeways.example.com goferd[4992]: [INFO][MainThread] gofer.agent.main:92 - agent started. 
```
What do you think should happen here?

jortel commented:
How did you install it?

jturel commented:
I adapted your install steps:

```
python setup.py install --no-compile --install-purelib /usr/lib/python2.7/site-packages
```

jturel commented:
Ok, I tested the latest stuff against python3+dnf on fedora27 and it's all working well, My only question is that installing gofer in that env brought in python2-gofer by default. Should we make python3 the default?

jturel commented:
Testing the latest RPM in my centos7 env with python2, yum is nearly working.

```
[root@centos7 gofer]# yum list installed | grep gofer
gofer.noarch                     2.12-0.git.2.9d1a79c.fc27      @jortel-gofer   
python2-gofer.noarch             2.12-0.git.2.9d1a79c.fc27      @jortel-gofer
```

a dependency error:
```
[root@centos7 gofer]# yum install python2-gofer-proton
Loaded plugins: fastestmirror, product-id, search-disabled-repos, subscription-manager
Default_Organization_animals_zoo                                                                                                                                                                                                                                   | 2.1 kB  00:00:00     
Loading mirror speeds from cached hostfile
 * base: mirror.team-cymru.org
 * epel: ftp.linux.ncsu.edu
 * extras: mirror.vcu.edu
 * updates: mirrors.centos.webair.com
Resolving Dependencies
--> Running transaction check
---> Package python2-gofer-proton.noarch 0:2.12-0.git.2.9d1a79c.fc27 will be installed
--> Processing Dependency: python2-qpid-proton >= 0.9-5 for package: python2-gofer-proton-2.12-0.git.2.9d1a79c.fc27.noarch
--> Finished Dependency Resolution
Error: Package: python2-gofer-proton-2.12-0.git.2.9d1a79c.fc27.noarch (jortel-gofer)
           Requires: python2-qpid-proton >= 0.9-5
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```

Should it be `python-qpid-proton` ?

jortel commented:
EL7 does not include python3 so most of the packages will not have been repackaged as python3- annd python3-.  I'm not surprised that you need to install as python-qpid-proton.

jortel commented:
The `gofer` package has a dependency on python-gofer which should resolve to the default python for the platform.  This seems correct.

jortel commented:
Actually, for EL6/EL7 we'll probably need to use existing 2.11 (or 2.7.8 in Sat channel) version of gofer.  Even building the python3-compat version of gofer on EL6/7 may be more trouble than it's worth given differences in macros and availability of either python2/python3.

Perhaps for now, it would make sense to test on el7 using: https://copr.fedorainfracloud.org/coprs/jortel/gofer/repo/epel-7/jortel-gofer-epel-7.repo

Thoughts?

jturel commented:
I'm good with that! Let me test this code against gofer 2.11 on el7.

jturel commented:
Getting further which is no surprise - but I am hitting this while uninstalling a package:
```
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9506]: [INFO][worker-0] gofer.rmi.dispatcher:599 - call: Content.uninstall() sn=e9950094-b297-4060-a102-94bc7c632a51 data={u'task_id': u'2719bb02-1e38-4d48-ae72-6e79feab643e', u'consumer_id': u'bd7f8439-842d-439e-bce0
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 - Handler failed.
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9506]: Exception AttributeError: AttributeError("'LibYum' object has no attribute '_history'",) in <bound method LibYum.__del__ of <katello.agent.pulp.libyum.LibYum object at 0x7f68f0180bd0>> ignored
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 - Traceback (most recent call last):
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -   File "/usr/lib/python2.7/site-packages/katello/agent/pulp/dispatcher.py", line 138, in uninstall
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -     _report = handler.uninstall(units, dict(options))
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -   File "/usr/lib/python2.7/site-packages/katello/agent/pulp/handler.py", line 176, in uninstall
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -     details = package.uninstall(patterns)
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -   File "/usr/lib/python2.7/site-packages/katello/agent/pulp/libyum.py", line 262, in uninstall
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -     with LibYum() as lib:
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -   File "/usr/lib/python2.7/site-packages/katello/agent/pulp/libyum.py", line 341, in __init__
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 -     self.preconf.optparser = self.__parser
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9533]: [ERROR][worker-0] katello.agent.pulp.dispatcher:142 - AttributeError: 'LibYum' object has no attribute 'preconf'
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9506]: [INFO][worker-0] gofer.agent.rmi:144 - Request: e9950094-b297-4060-a102-94bc7c632a51 processed in: 1.151 (seconds)
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9506]: [INFO][worker-0] gofer.agent.rmi:190 - Request: e9950094-b297-4060-a102-94bc7c632a51, committed
Apr 11 18:57:50 centos7.strangeways.example.com goferd[9506]: [INFO][worker-0] gofer.messaging.adapter.proton.connection:131 - closed: proton+amqps://centos7-katello-3-5.strangeways.example.com:5647
```

jturel commented on src/katello/agent/pulp/test.py:
```diff
@@ -0,0 +1,66 @@
+from libdnf import *
```
I guess this file should be under ./test/ somewhere :)

... which leads me to testing! It is definitely concerning that there are hardly any here especially because it could create a kind of behavior where devs are less enthusiastic about writing new tests if there are none/few to begin with. Are there tests for the libs from which this code was derived that could be adjusted to work here or would testing this be all new code?

jortel replied:
The `AttributeError: 'LibYum' object has no attribute 'preconf'` has me puzzled at the moment.  The `__init__()` code is almost exactly the same as it has been in pulp.  Will dig into it.

As for the tests, we *could* cannibalize the tests from pulp but they mock everything and don't catch integration problems which I think poses the greatest risk.  But, might be better than nothing.

jturel commented:
Sounds like refactoring those tests may be more trouble than it's worth. I've opened an issue to add coverage from our side at a later point. I think it'll be a good opportunity for others to get familiar with this code, so there is that advantage!

http://projects.theforeman.org/issues/23268

jortel commented:
Fixed ` AttributeError: 'LibYum' object has no attribute 'preconf'` and tested package install/remote using test.py on Centos7.0

jturel commented:
@jortel works great now!

This is close to done, unless you envision anything from your side. There are a few .pyo files in the commit to clean up.

As far as testing, we don't have the unit tests running as part of PRs yet - so I will check on those. However, the build is failing due to some things we do test with docker. You can fix that by removing 'zhenech/katello-client:el5' from the Makefile since we are no longer supporting it in this version

jturel commented:
I was able to get the unit tests passing w/ `make test-python2`  with a patch like this: https://paste.fedoraproject.org/paste/OgHipWl6BP0Hxt06FV5Bqw

jortel commented:
Excellent.
I'm still working to get the gofer.spec updated and built in Copr for EL6/7 but copr is _dog'n_ the last few days.  Taking 4-5 hours to run builds.  The _rel-eng_ team wants to use the upstream spec for EL5/6/7 builds. 

I plan to apply your testing patch and squash the PR.

jortel commented:
@jturel, also the 2.12 beta candidate is built in Copr for EL6/EL7 and F26/27 and would really appreciate your testing with them.

jturel commented:
@jortel I'll give it all another run-through today!

jturel commented:
Testing the new builds of latest code here with gofer2.12 beta...

el7 client worked great w/ python2 as expected. on fedora27 with python3 everything is mostly working - except on a package update I got this error:

```
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[3277]: [INFO][worker-0] gofer.rmi.dispatcher:547 - call: Content.update() sn=2f847167-956e-417a-bf50-af5cb1eabb0d data={'consumer_id': '75590e10-43e8-4f54-9575-21fc05dfdbf1', 'task_id': 'da1ee721-2e36-46c7-b740-8409e
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [INFO][worker-0] dnf:346 - Last metadata expiration check: 0:00:00 ago on Tue 17 Apr 2018 02:02:22 AM UTC.
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 - Handler failed.
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 - TypeError: Expected a string or a unicode object
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 - The above exception was the direct cause of the following exception:
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 - Traceback (most recent call last):
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -   File "/usr/lib/python3.6/site-packages/katello/agent/pulp/dispatcher.py", line 108, in update
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -     _report = handler.update(units, dict(options))
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -   File "/usr/lib/python3.6/site-packages/katello/agent/pulp/handler.py", line 152, in update
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -     details = package.update(patterns)
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -   File "/usr/lib/python3.6/site-packages/katello/agent/pulp/libdnf.py", line 150, in update
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -     lib.upgrade(patterns)
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -   File "/usr/lib/python3.6/site-packages/katello/agent/pulp/libdnf.py", line 334, in upgrade
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -     super(LibDnf, self).upgrade(p)
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -   File "/usr/lib/python3.6/site-packages/dnf/base.py", line 1764, in upgrade
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -     subj = dnf.subject.Subject(pkg_spec)
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -   File "/usr/lib/python3.6/site-packages/dnf/subject.py", line 36, in __init__
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -     self.subj = hawkey.Subject(pkg_spec)  # internal subject
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -   File "/usr/lib64/python3.6/site-packages/hawkey/__init__.py", line 428, in __init__
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 -     super(Subject, self).__init__(*args, **kwargs)
Apr 17 02:02:22 fedora27.strangeways.example.com goferd[7021]: [ERROR][worker-0] katello.agent.pulp.dispatcher:112 - SystemError: <method-wrapper '__init__' of Subject object at 0x7f0bdd0cd308> returned a result with an error set
```

Other than this (and the PR I opened against your branch) - all good

jortel commented:
Two outstanding things:
- The `map()` function changed in python3.  It's used by libyum to do loggging.  This needs to be fixed for compat.
- The libdnf needs to have logging (equal to yum) added.

jortel commented:
Ran this test using python3 on F27 and unable to reproduce so far.
```
[jortel@f27 ~]$ gofer rmi -u qpid+amqp://localhost -a demo -t Demo.install walrus-0.71-1 && rpm -qa|grep walrus
{u'resolved': [{u'name': u'walrus', u'qname': u'walrus-0.71-1.noarch', u'epoch': 0, u'version': u'0.71', u'release': u'1', u'arch': u'noarch', u'repoid': u'zoo'}], u'failed': [], u'deps': []}
walrus-0.71-1.noarch

[jortel@f27 ~]$ gofer rmi -u qpid+amqp://localhost -a demo -t Demo.update walrus && rpm -qa|grep walrus
{u'resolved': [{u'name': u'walrus', u'qname': u'walrus-5.21-1.noarch', u'epoch': 0, u'version': u'5.21', u'release': u'1', u'arch': u'noarch', u'repoid': u'zoo'}], u'failed': [], u'deps': []}
walrus-5.21-1.noarch

[jortel@f27 ~]$ gofer rmi -u qpid+amqp://localhost -a demo -t Demo.uninstall walrus && rpm -qa|grep walrus
{u'resolved': [{u'name': u'walrus', u'qname': u'walrus-5.21-1.noarch', u'epoch': 0, u'version': u'5.21', u'release': u'1', u'arch': u'noarch', u'repoid': u'@System'}], u'failed': [], u'deps': []}
```

jturel commented:
Will try these commands on my local f27

jturel commented:
@jortel 

Haven't tried your above steps yet but I was able to create a kind of base case to see the problem that you can test easily on your box:

```
[root@fedora27 katello-host-tools]# python3
Python 3.6.4 (default, Feb  1 2018, 11:06:09) 
[GCC 7.2.1 20170915 (Red Hat 7.2.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import hawkey
>>> from katello.agent.pulp.libdnf import Pattern
>>> p = Pattern({'name': 'walrus'})
>>> hawkey.Subject(p)
TypeError: Expected a string or a unicode object

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.6/site-packages/hawkey/__init__.py", line 428, in __init__
    super(Subject, self).__init__(*args, **kwargs)
SystemError: <method-wrapper '__init__' of Subject object at 0x7f6600ae4588> returned a result with an error set
>>> 

```

jortel commented:
Thanks, @jturel!!  Your reproducer gave me the hint I needed.  My `test.py` was not passing `Pattern` and the `Package.update()` was missing a `str()` on the pattern object.  Fixed.

Also, I updated the `Pattern` classes to improve testability.

jturel commented:
This is all working great! I ran through my tests again on centos6/yum/py2 and fedora27/dnf/py3 with gofer-2.12-0.git.6.14f0ee9

I tested installing, upgrading, removing packages and similar for package groups.

So close to merging.  How about https://github.com/jortel/katello-host-tools/pull/2 to fix travis

jortel commented:
Great, thanks @jturel.  I merged the travis PR.  Do you want me to squash again?  Or, looks like github has a squash option on the [merge] button.  Let me know.

jturel commented:
@jortel Hmm, the build got further but you can see the install step is failing with a syntax error on the el6 container:
```
SyntaxError: ('invalid syntax', ('src/katello/agent/pulp/libdnf.py', 359, 33, '        self.ids = {s.upper() for s in ids}\n'))
```

Did we say that we weren't supporting el6 either with this new version? If so, then another change similar to mine will need to be made in order to remove the el6 container from the Makefile test targets

jturel commented:
The container failing the install has this:

```
[root@aa5af60924a8 /]# cat /etc/redhat-release 
CentOS release 6.9 (Final)
[root@aa5af60924a8 /]# rpm -qa python
python-2.6.6-66.el6_8.x86_64
```
